design-documents/frontend/admin-guidelines.md

```diff
@@ -0,0 +1,194 @@
+
+### version
+
+The version of this `package.json` follows [semver](https://semver.org/) and will change with this document accordingly.
+
+```json
+"version": "0.1.0",
+```
+
+### scripts
+
+Customize your CRA configuration with [craco][].
+
+- [Don't eject](https://create-react-app.dev/docs/alternatives-to-ejecting).
+- Avoid forking, because you can oftentimes do the same thing by creating your own [craco][] plugin. See [craco plugins on npm](https://www.npmjs.com/search?q=craco%20plugin) and the [craco recipes](https://github.com/sharegate/craco/tree/master/recipes) for guidance.
+
+```json
+"scripts": {
+  "start": "craco start",
+  "build": "craco build",
+  "test": "craco test"
+},
+```
+
+### [browserslist](https://github.com/browserslist/browserslist#readme)
+
+Don't forget your [polyfills](https://create-react-app.dev/docs/supported-browsers-features#supported-language-features)!
+
+```json
+"browserslist": [
+  ">0.2%",
+  "not dead",
+  "not ie <= 11",
+  "not op_mini all"
+],
+```
+
+### dependencies
+
+#### Utility functions
+
+Use vanilla (native) JS if possible, utility functions if not. [You don't (may not) need Lodash/Underscore](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#you-dont-may-not-need-lodashunderscore-). See also [just](https://github.com/angus-c/just#just).
+
+```json
+"@queso/camel-case": "^0",
+"@queso/kebab-case": "^1",
+```
+
+#### SSR
+
+Prepare for [SSR](https://medium.com/walmartlabs/the-benefits-of-server-side-rendering-over-client-side-rendering-5d07ff2cefe8) by using [universal fetching](https://www.npmjs.com/package/ky-universal). Don't use [axios](https://www.npmjs.com/package/axios) or [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) directly.
+
+```json
+"ky": "^0",
+"ky-universal": "^0",
+```
+
+#### React
+
+```json
+"react": "^16",
+"react-dom": "^16",
+"react-helmet": "^5",
+```
+
+#### Routing
+
+See [The Future of React Router and @reach/router](https://reacttraining.com/blog/reach-react-router-future/). Basically, the new API will be much closer to `@reach/router` than React Router, so this prepares us for that future.
+
+```json
+"@reach/router": "^1",
+```
+
+#### State management
+
+[Immutability Helper](https://www.npmjs.com/package/immutability-helper) was [rewritten in TypeScript](https://github.com/kolodny/immutability-helper/pull/123) and strikes a balance between features and size.
+
+```json
+"immutability-helper": "^3",
+"react-redux": "^7",
+"redux": "^4",
+"redux-thunk": "^2",
+```
+
+#### Code splitting
+
+For context, read [Optimize your React application with Loadable Components](https://www.smooth-code.com/articles/code-splitting-react-loadable-components) and don't use dynamic imports in your components.
+
+```json
+"@loadable/component": "^5",
+```
+
+#### i18n
+
+```json
+"react-intl": "^3",
+```
+
+#### Theming
+
+```json
+"react-theme-context": "^2",
+```
+
+#### Relay
+
+```json
+"babel-plugin-relay": "^5",
+"react-relay": "^5",
```

Comment:
I'll just start the Relay vs Apollo discussion here...
With Apollo's adoption of [React hooks](https://www.npmjs.com/package/@apollo/react-hooks) in its latest release, it seems like Apollo will continue to evolve alongside React.

Reply:
I was definitely torn about this particular decision, but I knew it would be one or the other. I don't know enough about the internals, but my concern with Apollo is that (if I'm not mistaken) their strategy for consolidating queries and optimizing the payload to the user wasn't as efficient and opinionated as Relay. I'd be glad to be proven wrong on this point. Ultimately, I want the _user_ to have the best experience and performance possible!

I will say, however, that the Relay documentation is pretty bad compared to Apollo.

Reply:
Referencing the conclusion in [this article](https://www.prisma.io/blog/relay-vs-apollo-comparing-graphql-clients-for-react-apps-b40af58c1534#conclusion):

> Apollo optimizes for ease of use and flexibility, thus allowing developers to get started quickly with their applications. However, making sure that the local cache is kept consistent with Apollo requires lots of manual effort and, especially in larger applications, can get pretty complicated.

> Relay on the other hand is a very sophisticated system that is designed from the ground up to maximize efficiency in the communication with the server. React components express their data requirements using GraphQL fragments that are co-located to each component. This makes it easier to ensure that each component gets the data it needs even when data requirements for other components in the hierarchy are changing. Relay is doing a lot of work behind the scenes to keep its local store consistent with the state on the server. The power of Relay comes at the price difficult to grasp complexity and a slow learning curve.

Reply:
That's why I highlighted the article's date. I saw the F8 talk, which is what piqued my interest in Relay. I can confirm that the DX with Relay is not at all friendly, so if Apollo is better documented, it might be the right choice. Again, I was torn on this decision. It wasn't easy for me to find any fresh comparisons between what we saw at F8 and what Apollo currently offers. I'd really like to know the feature parity between them and the trajectories. I'm not sure that it matters though. Take a look at this!

<img width="391" alt="Screen Shot 2019-08-23 at 1 36 31 PM" src="https://user-images.githubusercontent.com/1058243/63615481-15f57c80-c5ab-11e9-98ea-340fe21d77ad.png">

I'm pretty sure that spike lines up with the F8 talk and tells us pretty well how the community received it. Now compare that to the `apollo-client`:

<img width="387" alt="Screen Shot 2019-08-23 at 1 36 12 PM" src="https://user-images.githubusercontent.com/1058243/63615541-3b828600-c5ab-11e9-9ab9-e52fdb66b017.png">

## Conclusion

I'm 100% OK changing this to the Apollo Client. Thanks for challenging this one!

Reply:
Addressed in https://github.com/magento/architecture/pull/231/commits/b1c19bc24ff6a13764b7aa3d7e9a2c64daaebb16